content/blog/2020-04-dplyr-1-0-0-rowwise/index.Rmarkdown

```diff
@@ -0,0 +1,205 @@
+
+Let's make those ideas concrete by creating a data frame a list-column. A little later, we'll come back to how you might actually get a list-column in a more realistic situation. The following data frame uses list columns to store things that would otherwise be challenging:
+
+* `x` contains vectors of different lengths.
+* `y` contains vectors of different types
+* `z` contains functions, which can't usually live in a data frame.
+
+```{r}
+df <- tibble(
+  x = list(1, 2:3, 4:6),
+  y = list(TRUE, 1, "a"),
+  z = list(sum, mean, sd)
+)
+df
+```
+
+When you have list-columns in a row-wise data frame, you can easily compute with each element of the list:
+
+```{r}
+df %>% 
+  rowwise() %>% 
+  summarise(
+    x_length = length(x),
+    y_type = typeof(y),
+    z_call = z(1:5)
+  )
+```
+
+This makes a row-wise `mutate()` or `summarise()` a general vectorisation tool, in the same way as the apply family in base R or the map family in purrr. It's now much simpler to solve a number of problems where we previously recommended learning about `map()`, `map2()`, `pmap()` and friends.
+
+## Use cases
+
+To finish up, I wanted to show off a couple of use cases where I think `rowwise()` provides a really elegant solution: simulations and modelling.
+
+### Simulation
+
+The basic idea of using `rowwise()` to perform simulation is to store all your simulation paramters in a data frame:
+
+```{r}
+df <- tribble(
+  ~id, ~ n, ~ min, ~ max,
+    1,   3,     0,     1,
+    2,   2,    10,   100,
+    3,   2,   100,  1000,
+)
+```
+
+Then you can either generate a list-column containing the simulated values with `mutate()`:
+
+```{r}
+df %>%
+  rowwise(id) %>%
+  mutate(data = list(runif(n, min, max)))
+```
+
+Or take advantage of [`summarise()`'s new capabilities](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/) and return one element per row: 
+
+```{r}
+df %>%
+  rowwise(id) %>%
+  summarise(x = runif(n, min, max))
+```
+
+Note that `id` is preserved in the output here because we defined it as an identifier variable in the call to `rowwise()`.
+
+[`vignette("rowwise")`](https://dplyr.tidyverse.org/dev/articles/rowwise.html#repeated-function-calls-1) expands on this idea to show you can generate parameter grids and vary the random distribution used in each row.
+
+### Group-wise models
+
+The new `nest_by()` function works similarly to `group_by()` but instead of storing the grouping data as metadata, visibly changes the structure. Now we have three rows (one for each group), and we have a list-col, `data`, that stores the data for that group. Also note that the output is a `rowwise()`; this is important because it’s going to make working with that list of data frames much easier.
```

Comment:
```suggestion
The new `nest_by()` function works similarly to `group_by()` but instead of storing the grouping data as metadata, visibly changes the structure. Now we have three rows (one for each group), and we have a list-col, `data`, that stores the data for that group. Also note that the output is a `rowwise()` object; this is important because it’s going to make working with that list of data frames much easier.
```

Reply:
You can obviously use a different word, but I think what you're going for here is the rowwise equivalent of "a grouped data frame," so you need _some_ noun (unless you take out the a, and say "the output is `rowwise()`", in which case I don't think rowwise should be formatted as a function).

Reply:
Yeah, sounds good.